src/bin/cli/logs/compare.rs

```diff
@@ -0,0 +1,150 @@
+use catalyst_toolbox::logs::sentry;
+use catalyst_toolbox::logs::sentry::{LogCmpStats, RawLog, SentryFragmentLog};
+use chain_core::property::Fragment;
+use jcli_lib::utils::io;
+use jormungandr_lib::interfaces::{
+    load_persistent_fragments_logs_from_folder_path, PersistentFragmentLog,
+};
+use serde::de::DeserializeOwned;
+use std::path::PathBuf;
+use structopt::StructOpt;
+use url::Url;
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("api-token parameter must be provided if sentry-url is set")]
+    MissingTokenParameter,
+
+    #[error(transparent)]
+    SentryLogsError(#[from] sentry::Error),
+
+    #[error(transparent)]
+    IoError(#[from] std::io::Error),
+
+    #[error("error deserializing logs from: {path}")]
+    DeserializeError {
+        path: PathBuf,
+        source: serde_json::Error,
+    },
+}
+
+#[derive(StructOpt)]
+#[structopt(rename_all = "kebab-case")]
+pub struct Compare {
+    #[structopt(long, conflicts_with = "sentry-logs")]
+    sentry_url: Option<Url>,
+
+    #[structopt(long, conflicts_with = "sentry-logs")]
+    api_token: Option<String>,
+
+    #[structopt(long)]
+    sentry_logs: Option<PathBuf>,
+
+    #[structopt(long)]
+    permanent_logs: PathBuf,
+}
+
+impl Compare {
+    pub fn exec(self) -> Result<(), Error> {
+        let Self {
+            sentry_url,
+            api_token,
+            sentry_logs,
+            permanent_logs,
+        } = self;
+        let sentry_logs: Vec<RawLog> = if let Some(url) = sentry_url {
+            if let Some(token) = api_token {
+                load_sentry_logs_from_url(url, token)?
+            } else {
+                return Err(Error::MissingTokenParameter);
+            }
+        } else if let Some(path) = sentry_logs {
+            load_logs_from_file(path)?
+        } else {
+            unreachable!()
+        };
+
+        let sentry_logs_data: Vec<SentryFragmentLog> = sentry_logs
+            .iter()
+            .enumerate()
+            .filter_map(
+                |(i, raw_log)| match raw_log.get("message").map(|v| v.as_str()).flatten() {
```

Comment:
`flat_map` in place of `map(..).flatten()`?

Reply:
It is an `Option`. I could flat_map earlier but I would have to compose a larger piece I think.

Reply:
yeah sorry, I meant `and_then` (just a different name for a flatmap on `Option`/`Result`)